File: frontend/src/services/analyticsService.ts
import { 
  AnalyticsApiResponse, 
  ChartDataType, 
  StatsOverview, 
  TimeRange, 
  TrendData, 
  DistributionData, 
  ComparisonData,
  PredictionResult
} from '../types/dataAnalysis';
import api from './api'; // 导入配置好的axios实例

// 基础API路径 - 已移除，因为api实例已设置baseURL
// const API_BASE_URL = '/api';

/**
 * 数据分析服务
 * 提供各种数据分析相关的API请求函数
 */
export const analyticsService = {
  /**
   * 获取统计数据概览
   */
  async getStatsOverview(): Promise<StatsOverview> {
    try {
      const response = await api.get(`system/visualization/overview`);
      
      if (response.data.status === 'error') {
        throw new Error(response.data.message || '获取统计数据失败');
      }
      
      return response.data.data;
    } catch (error) {
      console.error('获取统计数据概览失败:', error);
      
      // 返回模拟数据（实际项目中应该使用更合适的错误处理方式）
      return {
        patientCount: 1248,
        weekPatientChange: 4.2,
        doctorCount: 86,
        totalDevices: 152,
        activeDeviceRate: 82,
        avgTrainingTime: 45,
        rehabilitationRate: 68,
        patientSatisfaction: 92
      };
    }
  },
  
  /**
   * 获取趋势数据
   * @param dataType 数据类型
   * @param timeRange 时间范围
   */
  async getTrendData(dataType: ChartDataType, timeRange: TimeRange): Promise<TrendData[]> {
    try {
      const response = await api.get(`analytics/trend`, {
        params: { dataType, timeRange }
      });
      
      if (response.data.status === 'error') {
        throw new Error(response.data.message || '获取趋势数据失败');
      }
      
      return response.data.data;
    } catch (error) {
      console.error(`获取${dataType}趋势数据失败:`, error);
      
      // 返回模拟数据
      return generateMockTrendData(dataType, timeRange);
    }
  },
  
  /**
   * 获取分布数据
   * @param dataType 数据类型
   * @param timeRange 时间范围
   */
  async getDistributionData(dataType: ChartDataType, timeRange: TimeRange): Promise<DistributionData[]> {
    try {
      const response = await api.get(`analytics/distribution`, {
        params: { dataType, timeRange }
      });
      
      if (response.data.status === 'error') {
        throw new Error(response.data.message || '获取分布数据失败');
      }
      
      return response.data.data;
    } catch (error) {
      console.error(`获取${dataType}分布数据失败:`, error);
      
      // 返回模拟数据
      return generateMockDistributionData(dataType);
    }
  },
  
  /**
   * 获取对比数据
   * @param dataType 数据类型
   * @param timeRange 时间范围
   * @param compareWith 对比时间段/类别
   */
  async getComparisonData(
    dataType: ChartDataType, 
    timeRange: TimeRange,
    compareWith: string
  ): Promise<ComparisonData[]> {
    try {
      const response = await api.get(`analytics/comparison`, {
        params: { dataType, timeRange, compareWith }
      });
      
      if (response.data.status === 'error') {
        throw new Error(response.data.message || '获取对比数据失败');
      }
      
      return response.data.data;
    } catch (error) {
      console.error(`获取${dataType}对比数据失败:`, error);
      
      // 返回模拟数据
      return generateMockComparisonData(dataType, compareWith);
    }
  },
  
  /**
   * 导出报表数据
   * @param dataType 数据类型
   * @param timeRange 时间范围
   * @param format 导出格式
   */
  async exportReport(
    dataType: ChartDataType, 
    timeRange: TimeRange,
    format: 'csv' | 'xlsx' | 'pdf'
  ): Promise<Blob> {
    try {
      const response = await api.get(`analytics/export`, {
        params: { dataType, timeRange, format },
        responseType: 'blob',
        headers: { 
          Accept: format === 'pdf' ? 'application/pdf' : 'application/octet-stream' 
        }
      });
      
      return response.data;
    } catch (error) {
      console.error(`导出${dataType}报表数据失败:`, error);
      throw error;
    }
  },

  /**
   * 获取高级预测分析数据
   * @param deviceId 设备ID
   * @param dataType 数据类型
   * @param days 分析最近多少天的数据
   * @param predictionDays 预测未来多少天的数据
   * @param method 预测方法
   * @param confidenceInterval 是否计算置信区间
   * @param interval 数据聚合间隔
   */
  async getAdvancedPrediction(
    deviceId: string,
    dataType: string,
    days: number = 30,
    predictionDays: number = 7,
    method: 'linear' | 'arima' | 'prophet' | 'ensemble' = 'linear',
    confidenceInterval: boolean = true,
    interval: 'hour' | 'day' | 'week' | 'month' = 'day'
  ): Promise<PredictionResult> {
    try {
      const params = new URLSearchParams({
        data_type: dataType,
        days: days.toString(),
        prediction_days: predictionDays.toString(),
        method,
        confidence_interval: confidenceInterval.toString(),
        interval
      });
      
      const response = await api.get(
        `device-analysis/predict-advanced/${deviceId}`,
        { params }
      );
      
      if (response.data.status === 'error') {
        throw new Error(response.data.message || '获取预测分析数据失败');
      }
      
      return response.data;
    } catch (error) {
      console.error(`获取预测分析数据失败:`, error);
      
      // 返回模拟数据
      return generateMockPredictionData(dataType, method, predictionDays, confidenceInterval);
    }
  }
};

// 生成模拟的趋势数据
function generateMockTrendData(dataType: ChartDataType, timeRange: TimeRange): TrendData[] {
  const result: TrendData[] = [];
  const daysCount = timeRange === 'day' ? 24 : // 按小时
                  timeRange === 'week' ? 7 : 
                  timeRange === 'month' ? 30 : 
                  timeRange === 'quarter' ? 90 : 
                  timeRange === 'year' ? 12 : 30; // 默认30天
  
  const now = new Date();
  const isMonthly = timeRange === 'year'; // 年度数据按月展示
  
  for (let i = 0; i < daysCount; i++) {
    const date = new Date();
    
    if (isMonthly) {
      // 设置为前N个月
      date.setMonth(now.getMonth() - (daysCount - 1) + i);
      date.setDate(1); // 月初
    } else if (timeRange === 'day') {
      // 设置为前N小时
      date.setHours(now.getHours() - (daysCount - 1) + i);
    } else {
      // 设置为前N天
      date.setDate(now.getDate() - (daysCount - 1) + i);
    }
    
    // 根据数据类型生成不同的随机值
    let value: number;
    switch (dataType) {
      case 'patient':
        // 患者数据的基线较高
        value = Math.floor(1000 + Math.random() * 500) + (i * 5);
        break;
      case 'doctor':
        // 医生工作量，缓慢增长
        value = Math.floor(50 + Math.random() * 30) + (i * 0.2);
        break;
      case 'device':
        // 设备使用量，波动较大
        value = Math.floor(100 + Math.random() * 100) + (i * Math.sin(i * 0.5) * 5);
        break;
      case 'rehabilitation':
        // 康复数据，稳步增长
        value = Math.floor(60 + Math.random() * 20) + (i * 0.8);
        break;
      default:
        value = Math.floor(100 + Math.random() * 100);
    }
    
    const formattedDate = isMonthly
      ? `${date.getFullYear()}-${(date.getMonth() + 1).toString().padStart(2, '0')}`
      : timeRange === 'day'
        ? `${date.getHours().toString().padStart(2, '0')}:00`
        : `${date.getFullYear()}-${(date.getMonth() + 1).toString().padStart(2, '0')}-${date.getDate().toString().padStart(2, '0')}`;
    
    result.push({
      date: formattedDate,
      value: Math.max(0, value) // 确保值不为负数
    });
  }
  
  return result;
}

// 生成模拟的分布数据
function generateMockDistributionData(dataType: ChartDataType): DistributionData[] {
  switch (dataType) {
    case 'patient':
      // 患者年龄分布
      return [
        { name: '0-18岁', value: 126, color: '#8884d8' },
        { name: '19-35岁', value: 284, color: '#82ca9d' },
        { name: '36-50岁', value: 324, color: '#ffc658' },
        { name: '51-65岁', value: 286, color: '#ff8042' },
        { name: '66岁以上', value: 228, color: '#0088FE' }
      ];
    case 'doctor':
      // 医生专业分布
      return [
        { name: '神经科', value: 18, color: '#8884d8' },
        { name: '骨科', value: 22, color: '#82ca9d' },
        { name: '康复科', value: 28, color: '#ffc658' },
        { name: '内科', value: 12, color: '#ff8042' },
        { name: '其他', value: 6, color: '#0088FE' }
      ];
    case 'device':
      // 设备类型分布
      return [
        { name: '康复训练设备', value: 68, color: '#8884d8' },
        { name: '监测设备', value: 42, color: '#82ca9d' },
        { name: '辅助设备', value: 24, color: '#ffc658' },
        { name: '评估设备', value: 18, color: '#ff8042' }
      ];
    case 'rehabilitation':
      // 康复类型分布
      return [
        { name: '运动功能', value: 42, color: '#8884d8' },
        { name: '言语功能', value: 18, color: '#82ca9d' },
        { name: '认知功能', value: 16, color: '#ffc658' },
        { name: '日常生活', value: 24, color: '#ff8042' }
      ];
    default:
      return [
        { name: '类别A', value: 30, color: '#8884d8' },
        { name: '类别B', value: 25, color: '#82ca9d' },
        { name: '类别C', value: 35, color: '#ffc658' },
        { name: '类别D', value: 10, color: '#ff8042' }
      ];
  }
}

// 生成模拟的对比数据
function generateMockComparisonData(dataType: ChartDataType, compareWith: string): ComparisonData[] {
  const categories = dataType === 'patient' ? ['新患者', '复诊患者', '住院患者', '出院患者'] :
                  dataType === 'doctor' ? ['普通门诊', '专家门诊', '手术', '查房'] :
                  dataType === 'device' ? ['康复训练', '评估测试', '日常监测', '急救使用'] :
                  dataType === 'rehabilitation' ? ['初级康复', '中级康复', '高级康复', '康复评估'] :
                  ['类别A', '类别B', '类别C', '类别D'];
  
  const periods = compareWith === 'lastPeriod' ? 
    { current: '本期', previous: '上期' } : 
    { current: '今年', previous: '去年' };
    
  return categories.map(category => {
    const currentValue = Math.floor(50 + Math.random() * 100);
    const previousValue = Math.floor(currentValue * (0.7 + Math.random() * 0.6)); // 70%-130%的变化
    
    return {
      category,
      [periods.current]: currentValue,
      [periods.previous]: previousValue
    };
  });
}

// 生成模拟的预测数据
function generateMockPredictionData(
  dataType: string,
  method: string,
  predictionDays: number,
  withConfidence: boolean
): PredictionResult {
  const now = new Date();
  const historyData = [];
  
  // 生成历史数据
  for (let i = 0; i < 30; i++) {
    const date = new Date();
    date.setDate(now.getDate() - (30 - i));
    
    // 为不同数据类型生成不同范围的模拟值
    let baseValue = 0;
    switch (dataType) {
      case 'heart_rate':
        baseValue = 70 + Math.random() * 10;
        break;
      case 'blood_pressure':
        baseValue = 120 + Math.random() * 20;
        break;
      case 'temperature':
        baseValue = 36.5 + Math.random() * 1;
        break;
      case 'steps':
        baseValue = 5000 + Math.random() * 3000;
        break;
      default:
        baseValue = 100 + Math.random() * 50;
    }
    
    historyData.push({
      value: baseValue + i * 0.5 * (Math.random() > 0.5 ? 1 : -1),
      timestamp: date.toISOString(),
      data_type: dataType
    });
  }
  
  // 生成预测数据
  const predictions = [];
  const confidenceIntervals = [];
  
  let lastValue = historyData[historyData.length - 1].value;
  const trend = (lastValue - historyData[0].value) / 30;
  
  for (let i = 0; i < predictionDays; i++) {
    // 添加一些随机波动
    const randomFactor = Math.random() * 0.1 * lastValue;
    const predictedValue = lastValue + trend * (i + 1) + (randomFactor - (0.05 * lastValue));
    
    predictions.push(predictedValue);
    
    if (withConfidence) {
      // 增加随时间的不确定性
      const uncertainty = (i + 1) * 0.02 * lastValue;
      confidenceIntervals.push({
        lower: predictedValue - uncertainty,
        upper: predictedValue + uncertainty
      });
    }
  }
  
  return {
    status: 'success',
    data_type: dataType,
    device_id: 'mock-device-123',
    days_analyzed: 30,
    prediction_days: predictionDays,
    method: method,
    data_points: historyData.length,
    prediction: {
      values: predictions,
      confidence_interval: withConfidence ? confidenceIntervals : null
    },
    all_predictions: method === 'ensemble' ? {
      linear: {
        values: predictions.map(v => v * 0.9),
        confidence_interval: withConfidence ? confidenceIntervals.map(ci => ({
          lower: ci.lower * 0.9,
          upper: ci.upper * 0.9
        })) : null
      },
      arima: {
        values: predictions.map(v => v * 1.1),
        confidence_interval: withConfidence ? confidenceIntervals.map(ci => ({
          lower: ci.lower * 1.1,
          upper: ci.upper * 1.1
        })) : null
      },
      prophet: {
        values: predictions.map(v => v * 1.05),
        confidence_interval: withConfidence ? confidenceIntervals.map(ci => ({
          lower: ci.lower * 1.05,
          upper: ci.upper * 1.05
        })) : null
      },
      ensemble: {
        values: predictions,
        confidence_interval: withConfidence ? confidenceIntervals : null
      }
    } : null,
    history: {
      original: historyData.map(item => item.value),
      aggregated: historyData.map(item => ({
        date: item.timestamp.split('T')[0],
        value: item.value
      }))
    }
  };
} 

File: backend/app/services/analytics_service.py
from typing import Dict, List, Any, Optional
from datetime import datetime, timedelta
import random
from pymongo.database import Database
from bson import ObjectId

class AnalyticsService:
    """数据分析和可视化服务"""
    
    def __init__(self, db: Database):
        """初始化数据分析服务
        
        Args:
            db: MongoDB数据库连接
        """
        self.db = db
        self.patients_collection = db.patients
        self.doctors_collection = db.doctors
        self.devices_collection = db.devices
        self.rehabilitation_collection = db.rehabilitation_records
    
    async def get_stats_overview(self) -> Dict[str, Any]:
        """获取系统统计数据概览
        
        Returns:
            包含各种统计数据的字典
        """
        try:
            # 计算患者数量
            patient_count = await self.patients_collection.count_documents({})
            
            # 计算医生数量
            doctor_count = await self.doctors_collection.count_documents({})
            
            # 计算设备数量
            device_count = await self.devices_collection.count_documents({})
            
            # 计算活跃设备比例
            active_devices = await self.devices_collection.count_documents({"status": "active"})
            active_device_rate = (active_devices / device_count * 100) if device_count > 0 else 0
            
            # 计算一周内的患者增长率
            one_week_ago = datetime.utcnow() - timedelta(days=7)
            previous_patient_count = await self.patients_collection.count_documents(
                {"created_at": {"$lt": one_week_ago}}
            )
            week_patient_change = ((patient_count - previous_patient_count) / previous_patient_count * 100) if previous_patient_count > 0 else 0
            
            # 计算平均训练时间（分钟）
            rehab_records = self.rehabilitation_collection.find({})
            total_duration = 0
            record_count = 0
            
            async for record in rehab_records:
                if "duration" in record:
                    total_duration += record.get("duration", 0)
                    record_count += 1
            
            avg_training_time = total_duration / record_count if record_count > 0 else 0
            
            # 由于缺少实际数据，以下数据为模拟值
            rehabilitation_rate = 68  # 康复率
            patient_satisfaction = 92  # 患者满意度
            
            return {
                "patientCount": patient_count,
                "weekPatientChange": round(week_patient_change, 1),
                "doctorCount": doctor_count,
                "totalDevices": device_count,
                "activeDeviceRate": round(active_device_rate, 1),
                "avgTrainingTime": round(avg_training_time, 1),
                "rehabilitationRate": rehabilitation_rate,
                "patientSatisfaction": patient_satisfaction
            }
        except Exception as e:
            print(f"获取统计数据概览失败: {str(e)}")
            # 返回模拟数据
            return {
                "patientCount": 1248,
                "weekPatientChange": 4.2,
                "doctorCount": 86, 
                "totalDevices": 152,
                "activeDeviceRate": 82,
                "avgTrainingTime": 45,
                "rehabilitationRate": 68,
                "patientSatisfaction": 92
            }

    async def get_trend_data(self, data_type: str, time_range: str) -> List[Dict[str, Any]]:
        """获取趋势数据
        
        Args:
            data_type: 数据类型 (patient, doctor, device, rehabilitation)
            time_range: 时间范围 (day, week, month, quarter, year)
            
        Returns:
            趋势数据列表
        """
        try:
            # 实现实际的数据查询逻辑
            # 这里为了演示，返回模拟数据
            days_count = {
                "day": 24,  # 一天24小时
                "week": 7,  # 一周7天
                "month": 30,  # 一个月约30天
                "quarter": 90,  # 一个季度约90天
                "year": 12  # 一年12个月
            }.get(time_range, 30)
            
            result = []
            now = datetime.utcnow()
            is_monthly = time_range == "year"  # 年度数据按月展示
            
            for i in range(days_count):
                date = datetime.utcnow()
                
                if is_monthly:
                    # 设置为前N个月
                    date = datetime(now.year, now.month - (days_count - 1) + i, 1) if now.month > (days_count - 1) - i else datetime(now.year - 1, 12 - ((days_count - 1) - i - now.month), 1)
                elif time_range == "day":
                    # 按小时设置
                    date = now - timedelta(hours=(days_count - 1) - i)
                else:
                    # 按天设置
                    date = now - timedelta(days=(days_count - 1) - i)
                
                # 生成随机数据
                if data_type == "patient":
                    value = random.randint(10, 50)
                elif data_type == "doctor":
                    value = random.randint(1, 10)
                elif data_type == "device":
                    value = random.randint(5, 30)
                else:
                    value = random.randint(20, 100)
                
                result.append({
                    "timestamp": date.isoformat(),
                    "value": value
                })
            
            return result
        except Exception as e:
            print(f"获取趋势数据失败: {str(e)}")
            # 返回空列表
            return []
    
    async def get_distribution_data(self, data_type: str, time_range: str) -> List[Dict[str, Any]]:
        """获取分布数据
        
        Args:
            data_type: 数据类型 (patient, doctor, device, rehabilitation)
            time_range: 时间范围 (day, week, month, quarter, year)
            
        Returns:
            分布数据列表
        """
        try:
            # 实现实际的数据查询逻辑
            # 这里为了演示，返回模拟数据
            result = []
            
            if data_type == "patient":
                categories = ["老年患者", "中年患者", "青年患者", "儿童患者"]
            elif data_type == "doctor":
                categories = ["康复科", "骨科", "神经科", "内科", "外科"]
            elif data_type == "device":
                categories = ["上肢康复", "下肢康复", "平衡训练", "认知训练", "日常生活训练"]
            else:
                categories = ["完全康复", "明显改善", "轻微改善", "无变化", "恶化"]
            
            total_value = 100
            remaining = total_value
            
            for i, category in enumerate(categories):
                # 最后一个类别获取剩余值，确保总和为100
                if i == len(categories) - 1:
                    value = remaining
                else:
                    value = random.randint(5, remaining - (len(categories) - i - 1) * 5)
                    remaining -= value
                
                result.append({
                    "category": category,
                    "value": value
                })
            
            return result
        except Exception as e:
            print(f"获取分布数据失败: {str(e)}")
            # 返回空列表
            return []
    
    async def get_comparison_data(self, data_type: str, time_range: str, compare_with: str) -> List[Dict[str, Any]]:
        """获取对比数据
        
        Args:
            data_type: 数据类型 (patient, doctor, device, rehabilitation)
            time_range: 时间范围 (day, week, month, quarter, year)
            compare_with: 对比对象 (lastPeriod, lastYear, target)
            
        Returns:
            对比数据列表
        """
        try:
            # 实现实际的数据查询逻辑
            # 这里为了演示，返回模拟数据
            categories = ["接诊量", "康复率", "满意度", "随访率", "治疗时长"]
            result = []
            
            for category in categories:
                current_value = random.randint(50, 100)
                
                if compare_with == "lastPeriod":
                    compare_value = current_value * random.uniform(0.8, 1.2)
                elif compare_with == "lastYear":
                    compare_value = current_value * random.uniform(0.7, 1.3)
                else:  # target
                    compare_value = 100  # 目标值
                
                result.append({
                    "category": category,
                    "currentValue": round(current_value, 1),
                    "compareValue": round(compare_value, 1)
                })
            
            return result
        except Exception as e:
            print(f"获取对比数据失败: {str(e)}")
            # 返回空列表
            return [] 